rfd/0078-login-rules.md

```diff
@@ -0,0 +1,704 @@
+  #
+  # Login rules with lower numbered priorities will be applied first, followed
+  # by rules with priorities in increasing order. In case of a tie, login rules
+  # with the same priority will be ordered by a lexicographical sort of their
+  # names.
+  priority: 0
+
+  # traits_map will determine the traits of all users who log in to this cluster.
+  #
+  # This is a YAML map where the key must be a static string which will be the
+  # final trait key, and the value is a list of predicate expressions which each
+  # must return a set of strings. The final trait will be set to the union of
+  # the resulting string sets of all predicate expressions for that trait key.
+  #
+  # traits_map must contain the complete set of desired traits, any external
+  # traits not found here will not be included in the user's certificates.
+  traits_map:
+    # keep an external trait, unmodified
+    logins:
+      - external.logins
+
+    # rename an external trait
+    db_logins:
+      - external.Database_Usernames
+
+    # merge external traits
+    kube_groups:
+      - external.groups
+      - external.kubernetes_groups
+
+    # transform an external trait
+    apps:
+      - 'lower(external.apps)'
+
+    # extend an external trait
+    windows_logins:
+      - external.windows_logins
+      - "bill"
+
+    # conditionally extend an external trait
+    groups:
+      - 'ifelse(external.groups.contains("splunk"), external.groups.add("dbs"), external.groups)'
+
+    # static string values
+    tags:
+      # if the "expression" is a static string or returns a single strings, this
+      # will be automatically treated as a set of size 1.
+      - "teleport"
+      - "access"
+
+  # traits_expression is a single predicate expression which must return a dict which will
+  # set the user's traits during login.
+  #
+  # The dict must always have keys of type string and values of type set of
+  # strings, or an error will be returned when the expression is evaluated.
+  #
+  # traits_expression is an alternative to traits_map, only one or the other
+  # can be specified, never both in the same login_rule.
+  #
+  # traits_expression: >
+  #   external.remove("irrelevant", "internal", "tags")
+  #     .put("groups", 
+  #       ifelse(external.groups.contains("splunk"),
+  #         external.groups.add("dbs"),
+  #         external.groups))
+  #     .put("logins",
+  #       set(
+  #         "ubuntu",
+  #         ifelse(external.groups.contains("admins"), "root", ""),
+  #         ifelse(external.organization.contains("teleport"), "teleporters", "external"))
+  #       .remove(""))
+
+  # eventually login_rules could also return the desired teleport roles for the
+  # user to allow custom logic, but this will be out of scope for the initial
+  # RFD and implementation.
+  # roles: >
+  #   set("ssh-users", "db-users")
+```
+
+### Why both `traits_map` and `traits_expression`
+
+`traits_expression` is meant to be an extremely powerful way for users with
```

Comment:
style nit from "On writing well". If you remove "meant to be an extremely powerful" and "be able to" the sentence will still work:

`traits_expression` is a way for users with complex requirements to express their traits. It has the ...

Reply:
Yeah that is much more succinct, I like it. Thanks for the quick review!